File: leetcode/src/coin-change.cpp
/*
Top-Down DP
*/
#include <climits>
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {

        int coins_needed[10010];
        bool visited[10010];
        fill(coins_needed, coins_needed+amount+3, SHRT_MAX);
        fill(visited, visited+amount+3, false);
        coins_needed[0] = 0;
        visited[0] = true;

        auto helper = [&coins, &coins_needed, &visited](int curr_amt, auto& helper_ref) -> int {
            if (visited[curr_amt])
                return coins_needed[curr_amt];
            
            for (auto coin: coins) {
                if (curr_amt >= coin)
                    coins_needed[curr_amt] 
                            = min(coins_needed[curr_amt], helper_ref(curr_amt - coin, helper_ref)+1);
            }

            visited[curr_amt] = true;
            return coins_needed[curr_amt];

        };

        int result = helper(amount, helper);
        return result == SHRT_MAX ? -1 : result;
    }
};

/*
BOTTOM-UP DP
*/
#include <climits>
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int coins_needed[10010];
        fill(coins_needed, coins_needed+amount+1, INT_MAX);
        coins_needed[0] = 0;

        for (int i = 1; i <= amount; i ++) {
            for (auto coin: coins)
                if (i >= coin && coins_needed[i-coin] != INT_MAX)
                    coins_needed[i] = min(coins_needed[i], coins_needed[i-coin]+1);
        }


        return coins_needed[amount] == INT_MAX ? -1 : coins_needed[amount];
    }
};

File: leetcode/src/coin-change.py
"""
TOP_DOWN_DP:
"""

INT_MAX = 100000
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = dict()

        def helper(amt):
            if amt == 0:
                return 0
            if amt not in dp:
                dp[amt] = INT_MAX
                for coin in coins:
                    if (amt >= coin):
                        dp[amt] = min(dp[amt], helper(amt-coin)+1)
            return dp[amt]

        result = helper(amount)
        if (result == INT_MAX):
            return -1
        else:
            return result

